README
NAME
    Crypt::SSSS - implementation of Shamir's Secret Sharing System.

SYNOPSIS
        use Crypt::SSSS;

        # use (3, 3) scheme
        my $shares = ssss_distribute(
            message => "\x06\x1c\x08",
            k       => 3,
        );

        # Save shares
        for my $share (1 .. 3) {
            open my $fh, '>', "share${share}.dat";
            print $fh $shares->{$share}->binary;
            close $fh;
        }

        # Reconstruct message
        my $ishares = {};
        for my $share (1 .. 3) {
            open my $fh, '<', "share${share}.dat";
            $ishares->{$share} = do {
                local $/;    # slurp!
                <$fh>;
            };
            close $fh;
        }

        print "Original message: ", sprintf '"\x%02x\x%02x\x%02x"',
          unpack('C*', ssss_reconstruct(p => 257, shares => $ishares));

DESCRIPTION
    Implementation of Shamir's Secret Sharing Scheme.

ATTRIBUTES
    Crypt::SSSS implements the following attributes.

  "ssss_distribute"
        my $shares = ssss_distribute(
            message => $message,
            k       => $k,
            p       => $p,         # 257 by default
            n       => $n,         # By default equals to k
        );

    Distribute $message to $n shares, so that any $k shares would be enough
    to reconstruct the secret. $p is a prime number.

    Returns hashref of Crypt::SSSS::Message.

  "ssss_reconstruct"
        my $secret = ssss_reconstruct(
            shares => $shares,
            p      => $p,        # 257 by default
        );

    Reconstruct message from given $shares. $p is a prime number used to
    distribute message.

AUTHOR
    Sergey Zasenko, "undef@cpan.org".

CREDITS
      Mohammad S Anwar (Manwar)

COPYRIGHT AND LICENSE
    Copyright (C) 2011, 2016, Sergey Zasenko.

    This program is free software, you can redistribute it and/or modify it
    under the same terms as Perl 5.10.



Changes
This file documents the revision history for Perl extension Protocol::Redis

0.3  0000-00-00 00:00:00
        - README added to distribution (MANWAR)
        - Makefile.PL fixups (MANWAR, UNDEF)
        - Changes fixup

0.2  2016-01-08 00:00:00
        - fixed repository url (MANWAR)

0.1  2016.01-06 00:00:00
        - first release


lib/TSS.pm
package TSS;

use strict;
use warnings;

our $VERSION = 0.1;

use Math::BigInt  ();
use Math::Polynom ();
use POSIX qw(ceil pow);
use TSS::Message  ();

require Carp;

sub tss_distribute {
    my (%data) = @_;

    my $message = $data{message} or Carp::croak 'Missed "message" argument';

    my $k = $data{k} or Carp::croak 'Missed "k" argument';
    my $n = $data{n} || $k;

    my $p = $data{p} or Carp::croak 'Missed "p" argument';

    my $messages = {};

    for my $x (1 .. $n) {
        $messages->{$x} = TSS::Message->new(p => $p);
    }

    my @args = unpack 'C*', $message;
    while (@args) {
        my @a = splice @args, 0, $k;

        for my $x (1 .. $n) {

            my $res = 0;
            for my $pow (0 .. $k - 1) {
                $res += $a[$pow] * pow($x, $pow);

            }

            # print "$x → ", $res % $p, "\n";
            $messages->{$x}->push_data($res % $p);
        }
    }

    $messages;
}

sub tss_reconstruct {
    my ($k, $p, $messages, $size) = @_;

    Carp::croak "Need at least $k messages" if (keys %$messages < $k);


    my @xs = keys %$messages;

    my %mdata;
    foreach my $x (@xs) {
        $mdata{$x} =
          TSS::Message->build_from_binary($p, $messages->{$x})->get_data;
    }

    $size ||= @{(values %mdata)[0]};

    my $message = '';

    for (my $l = 0; $l < $size; $l++) {
        my $fx = Math::Polynom->new(0 => 0);
        for my $i (@xs) {
            my $pl = Math::Polynom->new(0 => 1);
            my $d = 1;
            for my $j (@xs) {
                if ($j != $i) {
                    $pl = $pl->multiply(Math::Polynom->new(1 => 1, 0 => -$j));
                    $d *= $i - $j;
                }
            }
            $d += $p if $d < 0;

            my ($m) = extended_gcb($d, $p);
            $m += $p if $m < 0;

            $fx = $fx->add($pl->multiply($m * $mdata{$i}->[$l]));
        }

        for (values %{$fx->{polynom}}) {
            $_ %= $p;
            $_ += $p if $_ < 0;
        }

        for (my $i = 0; $i < $k; $i++) {
            $message .= pack 'C', $fx->{polynom}->{$i};
        }
    }

    $message;
}

sub extended_gcb {
    my ($a, $b) = @_;

    return (1, 0) if $b == 0;

    my $q = int($a / $b);
    my $r = $a % $b;
    my ($s, $t) = extended_gcb($b, $r);

    return ($t, $s - $q * $t);
}

1;
__END__

=head1 NAME

TSS - Shamir's Threshold Sharing System implementation.

=head1 SYNOPSIS

    use TSS;

    # use (3, 3) scheme
    my $shares = TSS::tss_distribute(
        message => "\x06\x1c\x08",
        k       => 3,
        n       => 3,
        p       => 257);

    # Save shares
    for my $share (1..3) {
        open my $fh, '>', "share${share}.dat";
        print $fh $shares->{$share}->binary;
        close $fh;
    }

    # Reconstruct message
    my $ishares = {};
    for my $share (1..3) {
        open my $fh, '<', "share${share}.dat";
        $ishares->{$share} = do {
            local $/; # slurp!
            <$fh>;
        };
        close $fh;
    }

    print "Original message: ", sprintf '"\x%02x\x%02x\x%02x"',
        unpack('C*', TSS::tss_reconstruct(3, 257, $ishares));

=head1 AUTHOR

Sergey Zasenko, C<undef@cpan.org>.

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2011, Sergey Zasenko.

This program is free software, you can redistribute it and/or modify it under
the same terms as Perl 5.10.

=cut


lib/Crypt/SSSS/Message.pm
package Crypt::SSSS::Message;

use strict;
use warnings;

require Carp;

sub new {
    my $class = shift;
    my %args  = @_;

    Carp::croak('Missed "p" argument') unless $args{p};

    bless {
        _data => $args{data} || [],
        _p    => $args{p},
        _size => _determine_chunk_size($args{p} - 1)
    }, $class;
}

sub build_from_binary {
    my ($class, $p, $string) = @_;

    my $self = $class->new(p => $p);

    my $rdata = 0;
    my $rsize = 0;

    my @chunks = unpack 'C*', $string;

    my $size = $self->{_size};

    my $smask = 0;
    $smask = ($smask << 1) | 1 for (1 .. $size);

    while (@chunks > 0) {
        $rdata = $rdata << 8 | shift @chunks;
        $rsize += 8;

        while ($rsize >= $size) {
            $rsize -= $size;

            my $mask = $smask << $rsize;
            $self->push_data(($rdata & $mask) >> $rsize);
            $rdata &= ~$mask;
        }
    }

    $self;
}

sub push_data {
    my ($self, $data) = @_;

    Carp::croak('Data greater than p') if $data > $self->{_p};
    push @{$self->{_data}}, $data;
}

sub binary {
    my $self = shift;

    my $rsize = 0;
    my $rdata = 0x00;

    my $str;

    foreach my $lchunk (@{$self->{_data}}) {
        my $size = $self->{_size};

        my $chunk = $rdata << $size | $lchunk;
        $size += $rsize;

        while ($size > 8) {
            $size -= 8;
            my $mask = 0xff << $size;

            my $data = ($chunk & $mask) >> $size;
            $chunk &= ~$mask;

            $str .= pack 'C', $data;
        }
        $rsize = $size;
        $rdata = $chunk;
    }

    $str .= pack 'C', ($rdata << 8 - $rsize);

    $str;
}

sub get_data {
    my $self = shift;
    return $self->{_data};
}

sub get_p {
    my $self = shift;
    return $self->{_p};
}

# Hope we will not have p greater than dword can have
# Same as log2
sub _sig_bit {
    my $x = shift;

    for (my $i = 0; $i < 32; $i++) {
        return $i
          if $x < (1 << $i);
    }
    return;
}

sub _determine_chunk_size {
    my ($p) = @_;

    _sig_bit($p);
}


1;


lib/Crypt/SSSS.pm
package Crypt::SSSS;

use strict;
use warnings;

our $VERSION = 0.2;

require Exporter;

our @ISA    = qw(Exporter);
our @EXPORT = qw(ssss_distribute ssss_reconstruct);

use POSIX qw(ceil pow);
use Crypt::SSSS::Message;

require Carp;

sub ssss_distribute(%) {
    my (%data) = @_;

    my $message = $data{message} or Carp::croak 'Missed "message" argument';

    my $k = $data{k} or Carp::croak 'Missed "k" argument';
    my $n = $data{n} || $k;

    my $p = $data{p} || 257;

    my $shares = {};

    for my $x (1 .. $n) {
        $shares->{$x} = Crypt::SSSS::Message->new(p => $p);
    }

    my $chunks;
    if (my $ref = ref $message) {
        Carp::croak qw/"message" has unsupported type "$ref"/
          unless $ref eq 'Crypt::SSSS::Message';

        $chunks = $message->get_data;
    }
    else {
        $chunks = [unpack (($data{pack_size} || 'C') . '*', $message)];
    }
    while (@$chunks) {
        my @a = splice @$chunks, 0, $k;

        for my $x (1 .. $n) {

            my $res = 0;
            for my $pow (0 .. $k - 1) {
                $res += ($a[$pow] || 0) * pow($x, $pow);

            }

            # print "$x → ", $res % $p, "\n";
            $shares->{$x}->push_data($res % $p);
        }
    }

    $shares;
}

sub ssss_reconstruct(%) {
    my (%data) = @_;

    my $shares = $data{shares};
    my $p = $data{p} || '257';

    my @xs = keys %$shares;
    my $k = @xs;

    my %mdata;
    foreach my $x (@xs) {
        $mdata{$x} =
          Crypt::SSSS::Message->build_from_binary($p, $shares->{$x})
          ->get_data;
    }

    my $size = $data{size} || @{(values %mdata)[0]};

    my $message = '';

    my $pack_size = $data{pack_size} || 'C';

    for (my $l = 0; $l < $size; $l++) {
        my @fx = ();
        for my $i (@xs) {

            # Plynom
            my @pl = (1);

            # Divider
            my $d = 1;
            for my $j (@xs) {
                if ($j != $i) {

                    # Multiply polinoms
                    my @opl = @pl;
                    unshift @pl, 0;
                    for (my $i = 0; $i < @opl; $i++) {
                        $pl[$i] += -$j * $opl[$i];
                    }
                    $d *= $i - $j;
                }
            }
            $d += $p if $d < 0;

            my ($m) = extended_gcb($d, $p);
            $m += $p if $m < 0;

            while (@fx < @pl) {
                push @fx, 0;
            }

            # Add our polynom (multiplied by constant)
            for (my $j = 0; $j < @pl; $j++) {
                $fx[$j] += $m * $mdata{$i}->[$l] * $pl[$j];
            }
        }

        for (@fx) {
            $_ %= $p;
            $_ += $p if $_ < 0;
        }

        for (my $i = 0; $i < $k; $i++) {
            $message .= pack $pack_size, $fx[$i];
        }
    }

    $message;
}

sub extended_gcb {
    my ($a, $b) = @_;

    return (1, 0) if $b == 0;

    my $q = int($a / $b);
    my $r = $a % $b;
    my ($s, $t) = extended_gcb($b, $r);

    return ($t, $s - $q * $t);
}

1;
__END__

=head1 NAME

Crypt::SSSS - implementation of Shamir's Secret Sharing System.

=head1 SYNOPSIS

    use Crypt::SSSS;

    # use (3, 3) scheme
    my $shares = ssss_distribute(
        message => "\x06\x1c\x08",
        k       => 3,
    );

    # Save shares
    for my $share (1 .. 3) {
        open my $fh, '>', "share${share}.dat";
        print $fh $shares->{$share}->binary;
        close $fh;
    }

    # Reconstruct message
    my $ishares = {};
    for my $share (1 .. 3) {
        open my $fh, '<', "share${share}.dat";
        $ishares->{$share} = do {
            local $/;    # slurp!
            <$fh>;
        };
        close $fh;
    }

    print "Original message: ", sprintf '"\x%02x\x%02x\x%02x"',
      unpack('C*', ssss_reconstruct(p => 257, shares => $ishares));

=head1 DESCRIPTION

Implementation of Shamir's Secret Sharing Scheme.

=head1 ATTRIBUTES

Crypt::SSSS implements the following attributes.

=head2 C<ssss_distribute>

    my $shares = ssss_distribute(
        message => $message,
        k       => $k,
        p       => $p,         # 257 by default
        n       => $n,         # By default equals to k
    );

Distribute C<$message> to C<$n> shares, so that any C<$k> shares would be
enough to reconstruct the secret. C<$p> is a prime number.

Returns hashref of Crypt::SSSS::Message.

=head2 C<ssss_reconstruct>

    my $secret = ssss_reconstruct(
        shares => $shares,
        p      => $p,        # 257 by default
    );

Reconstruct message from given C<$shares>. C<$p> is a prime number used to
distribute message.

=head1 AUTHOR

Sergey Zasenko, C<undef@cpan.org>.

=head1 CREDITS

=over 2

=item Mohammad S Anwar (MANWAR)

=back

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2011, 2016, Sergey Zasenko.

This program is free software, you can redistribute it and/or modify it under
the same terms as Perl 5.10.

=cut
